### PR TITLE
Cache Active Validator Indices, Count, and Balances

### PR DESCRIPTION
**Pull request by terencechain**
Note: Sorry for the big diff, these were suppose to be simple map cache, but got good feedback to change them to fifo caches so we can guard against an upper bound. The functionalities are simple and easy to review

This PR adds caches to the following caches to avoid expensive operations such as looping over validator list (only need to do this once per epoch):

* **startShardCache** - tracks the start shard of a given epoch
* **totalBalanceCache** - tracks the total validator balance of a given epoch
* **totalActiveBalanceCache** - tracks the total active validator balance of a given epoch
* **activeCountCache** - tracks the active validator count of a given epoch
* **activeIndicesCache** - tracks the active validator indices of a given epoch (expensive)
* **epochSeedCache** - caches the epochSeed of a current epoch

It cuts process epoch time from **96s -> 0.5s** for 16324 validators and **5mins -> 2.5s** for 62500 validators

### CODE REVIEW

**Comment by terencechain** [comment]
Talked with @prestonvanloon, I'll add a max size to protect against upper bound.

File: beacon-chain/blockchain/block_processing_test.go [hunk 1]
```diff
@@ -935,3 +936,77 @@ func TestSaveValidatorIdx_IdxNotInState(t *testing.T) {
 		t.Error("Did not get wanted validator from activation queue")
 	}
 }
+
+func TestNewFinalizedBlock_CanClearCaches(t *testing.T) {
```
Comment (terencechain) [inline beacon-chain/blockchain/block_processing_test.go]: Add test to verify a new finalized block will clear cache

File: beacon-chain/cache/active_count.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	// ErrNotActiveCountInfo will be returned when a cache object is not a pointer to
+	// a ActiveCountByEpoch struct.
+	ErrNotActiveCountInfo = errors.New("object is not a active count obj")
+
+	// maxActiveCountListSize defines the max number of active count can cache.
+	maxActiveCountListSize = 1000
+
+	// Metrics.
+	activeCountCacheMiss = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "active_count_cache_miss",
```
Comment (prestonvanloon) [inline beacon-chain/cache/active_count.go]: From the metrics page, I think a reader would have a hard time understanding what this metric is. Can you add more to the name or help message?

File: beacon-chain/cache/active_indices.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	// a ActiveIndicesByEpoch struct.
+	ErrNotActiveIndicesInfo = errors.New("object is not a active indices list")
+
+	// maxActiveIndicesListSize defines the max number of active indices can cache.
+	maxActiveIndicesListSize = 4
+
+	// Metrics.
+	activeIndicesCacheMiss = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "active_indices_cache_miss",
+		Help: "The number of active indices requests that aren't present in the cache.",
```
Comment (prestonvanloon) [inline beacon-chain/cache/active_indices.go]: This too, please specify that these are validator indices

File: beacon-chain/cache/active_count.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/cache/active_count.go]: yeah good point, `active_validator_count_cache_miss` is better here

**Pull request merged**